### PR TITLE
Used type erased allocator API

### PR DESCRIPTION
**Pull request by apankratovantonp**
### Details:
 - Reduce users code connectivity with OpenVINO

### CODE REVIEW

File: src/core/include/openvino/runtime/allocator.hpp [hunk 1]
```diff
@@ -61,20 +65,61 @@ class Tensor;
...
+        explicit Impl(Args&&... args) : a(std::forward<Args>(args)...) {}
+        void* addressof() override {
+            return &a;
+        }
+        const std::type_info& type_info() const override {
+            return typeid(a);
+        }
+        void* allocate(const size_t bytes, const size_t alignment = alignof(max_align_t)) override {
+            return a.allocate(bytes, alignment);
+        };
```
Comment (nosovmik) [inline src/core/include/openvino/runtime/allocator.hpp:108]: `;` is not necessary

File: src/core/tests/ov_tensor_test.cpp [hunk 2]
```diff
@@ -51,22 +51,37 @@ TEST_F(OVTensorTest, operators) {
     ASSERT_TRUE(!t);
 }
 
-class OVMockAllocator : public ov::AllocatorImpl {
-public:
-    MOCK_METHOD(void*, allocate, (size_t, size_t), ());
-    MOCK_METHOD(void, deallocate, (void*, size_t, size_t), ());                  // NOLINT(readability/casting)
-    MOCK_METHOD(bool, is_equal, (const ov::AllocatorImpl&), (const, noexcept));  // NOLINT(readability/casting)
+struct OVMockAllocator {
```
Comment (nosovmik) [inline src/core/tests/ov_tensor_test.cpp:54]: should we keep old 'OVMockAllocator' test to ensure that 'deprecated' API still works for end user?

File: src/core/src/runtime/allocator.cpp [hunk 3]
```diff
@@ -10,20 +10,66 @@
 #include "openvino/core/except.hpp"
 
 namespace ov {
+struct DeafaultAllocator {
```
[the remark this answers is not in the document]
Comment (apankratovantonp) [reply src/core/src/runtime/allocator.cpp]: fixed

File: samples/cpp/benchmark_app/shared_tensor_allocator.hpp [hunk 4]
```diff
@@ -4,32 +4,35 @@
...
-#include "openvino/runtime/allocator.hpp"
-
-class SharedTensorAllocator : public ov::AllocatorImpl {
+class SharedTensorAllocator {
 public:
     SharedTensorAllocator(size_t sizeBytes) : size(sizeBytes) {
         data = new char[size];
     }
+    SharedTensorAllocator(const SharedTensorAllocator&) = delete;
+    SharedTensorAllocator(SharedTensorAllocator&& other) : data{other.data} {
```
[the remark this answers is not in the document]
Comment (apankratovantonp) [reply samples/cpp/benchmark_app/shared_tensor_allocator.hpp]: removed as unnecessary

File: src/core/include/openvino/runtime/allocator.hpp (hunk 1 above)
Comment (apankratovantonp) [reply src/core/include/openvino/runtime/allocator.hpp]: removed

File: src/core/tests/ov_tensor_test.cpp (hunk 2 above)
Comment (apankratovantonp) [reply src/core/tests/ov_tensor_test.cpp:74]: added

**Comment by akladiev** [comment]
This PR will be closed in 2 weeks in case of no activity.

**Comment by akladiev** [comment]
This PR was closed because it has been stalled for 2 week with no activity.

**Pull request closed without merging**